central/resourcecollection/datastore/datastore_impl.go

```diff
@@ -427,3 +431,156 @@ func verifyCollectionObjectNotEmpty(obj *storage.ResourceCollection) error {
 	}
 	return nil
 }
+
+func (ds *datastoreImpl) ResolveListDeployments(ctx context.Context, collection *storage.ResourceCollection) ([]*storage.ListDeployment, error) {
+
+	if err := verifyCollectionConstraints(collection); err != nil {
+		return nil, err
+	}
+
+	query, err := ds.resolveCollectionQuery(ctx, collection)
+	if err != nil {
+		return nil, err
+	}
+	return ds.deploymentDS.SearchListDeployments(ctx, query)
+}
+
+func (ds *datastoreImpl) resolveCollectionQuery(ctx context.Context, collection *storage.ResourceCollection) (*v1.Query, error) {
+	var collections []*storage.ResourceCollection
+	var collectionSet set.Set[string]
+	var disjunctions []*v1.Query
+
+	collections = append(collections, collection)
+
+	ds.lock.RLock()
```

Comment:
Because we have to expand the collection objects that are embedded by querying the datastore level for those objects, so we take a read lock out to ensure that the objects don't go stale or get removed halfway through the expansion. For the future I envision the main thread being a process of just resolving these objects recursively until we've resolved the whole tree of objects, and spawn go routines as we go to turn the objects into queries

Reply:
👍